protocols/relay/src/v2/relay/handler.rs

```diff
@@ -0,0 +1,892 @@
+                        self.pending_error = Some(ProtocolsHandlerUpgrErr::Upgrade(
+                            upgrade::UpgradeError::Apply(EitherError::B(error)),
+                        ));
+                        Status::ConnectionFailed
+                    }
+                    outbound_stop::UpgradeError::UnexpectedStatus(status) => {
+                        match status {
+                            Status::Ok => {
+                                unreachable!("Status success is explicitly exempt.")
+                            }
+                            // A destination node returning nonsensical status is a protocol
+                            // violation. Thus terminate the connection.
+                            Status::ReservationRefused
+                            | Status::NoReservation
+                            | Status::ConnectionFailed => {
+                                self.pending_error = Some(ProtocolsHandlerUpgrErr::Upgrade(
+                                    upgrade::UpgradeError::Apply(EitherError::B(error)),
+                                ));
+                            }
+                            // With either status below there is no reason to stay connected.
+                            // Thus terminate the connection.
+                            Status::MalformedMessage | Status::UnexpectedMessage => {
+                                self.pending_error = Some(ProtocolsHandlerUpgrErr::Upgrade(
+                                    upgrade::UpgradeError::Apply(EitherError::B(error)),
+                                ))
+                            }
+                            // While useless for reaching this particular destination, the
+                            // connection to the relay might still proof helpful for other
+                            // destinations. Thus do not terminate the connection.
+                            Status::ResourceLimitExceeded | Status::PermissionDenied => {}
+                        }
+                        status
+                    }
+                }
+            }
+        };
+
+        let OutboundOpenInfo {
+            circuit_id,
+            inbound_circuit_req,
+            src_peer_id,
+            src_connection_id,
+        } = open_info;
+
+        self.queued_events.push_back(ProtocolsHandlerEvent::Custom(
+            Event::OutboundConnectNegotiationFailed {
+                circuit_id,
+                src_peer_id,
+                src_connection_id,
+                inbound_circuit_req,
+                status,
+            },
+        ));
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        self.keep_alive
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ProtocolsHandlerEvent<
+            Self::OutboundProtocol,
+            Self::OutboundOpenInfo,
+            Self::OutEvent,
+            Self::Error,
+        >,
+    > {
```

Comment:
> You'd have to keep the boxed future around

Under the assumption that the `ProtocolsHandler` trait stays the same, that future would potentially have a mutable reference to itself (the handler), which would not compile due to being a self-referential structure.

> When should we use async/await syntax

As much as possible, especially for linear executions, e.g. as in `protocols/relay/v2/protocol/`.

Though when something needs additional input (e.g. see various `inject_*` methods), I don't see how to achieve this with `async/await`. We could use `async-trait`, rename `poll` to `next` and make it an `async` function. Problems I see is whether it is safe to drop (i.e. cancel) a future retrieved via `ProtocolsHandler::next`, or does one need to always drive it to completion?
  - Being able to cancel a `ProtocolsHandler::next` future would allow calling the various `inject_*` methods in between.
  - Implementing `ProtocolsHandler::next` in a way that it can be dropped at any point in time is very hard. One needs to expect to be canceled across every `.awat` point.

Does the above make sense @MarcoPolo? I am happy to write this up in greater detail. That would be time well spend. I have thought about this a lot, though thus far I have been unable to come up with a better solution.

See also https://blog.yoshuawuyts.com/async-cancellation-1/#an-async-trait-which-can-t-be-cancelled for an interesting read on the topic.

Reply:
Resolving here. Feel free to un-resolve in case you want to continue the discussion.